presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/ReorderJoins.java

```diff
@@ -147,38 +147,20 @@ public int compare(JoinEnumerationResult result1, JoinEnumerationResult result2)
             };
         }
 
-        private JoinEnumerationResult chooseJoinOrder(List<PlanNode> sources, List<Symbol> outputSymbols, boolean isCrossJoinPhase)
+        private JoinEnumerationResult chooseJoinOrder(List<PlanNode> sources, List<Symbol> outputSymbols)
         {
             Set<PlanNode> multiJoinKey = ImmutableSet.copyOf(sources);
             JoinEnumerationResult bestResult = memo.get(multiJoinKey);
             if (bestResult == null) {
                 checkState(sources.size() > 1, "sources size is less than or equal to one");
-                List<Optional<JoinEnumerationResult>> results = generatePartitions(sources.size())
-                        .map(partitioning -> createJoinAccordingToPartitioning(sources, outputSymbols, partitioning, isCrossJoinPhase))
-                        .collect(toImmutableList());
+                bestResult = generatePartitions(sources.size())
+                        .map(partitioning -> createJoinAccordingToPartitioning(sources, outputSymbols, partitioning))
+                        .filter(result -> !result.cost.isUnknown())
+                        .min(resultOrdering).orElse(UNKNOWN_COST_RESULT);
 
-                // If costs are unknown we can't compare them. We return a result with unknown cost.
```

Comment:
neither cuts off other computations because they all happen at once in the map phase of the stream.  I can change the whole thing to a for loop so we can break out of it.